text/0043-reproducible-builds.md

```diff
@@ -0,0 +1,55 @@
+# Expanding the Criteria for Reproducible Builds
+
+## Summary
+
+Given the same set of inputs and without leveraging caching, Paketo buildpacks
+should reproducibly build images.
+
+## Motivation
+
+The current set of Paketo buildpacks do a reasonably good job of helping
+developers to ship reproducible build artifacts. They reuse layers where
+possible, and don't modify files in ways that aren't reproducible. So, for
+users taking advantage of the caching behaviors of CNBs, the reproducible build
+story is largely solved.
+
+However, we can do better in some other cases. Specifically, given a codebase
+that is not changing, a set of buildpacks that are not changing, and stacks
+that are not changing, we should be able to reproduce the exact same image on
+subsequent runs without leveraging the caching features of CNB.
+
+Unfortunately, there are some cases where we've made choices that prevent this
+type of reproducible build. These cases should be removed or minimized to the
+greatest extent possible.
+
+## Detailed Explanation
+
+Many of the buildpacks contain behavior that results in non-reproducible images
+due to how they treat layer metadata. Specifically, they include a `built_at`
+metadata field that includes a timestamp for the instant that this layer was
+built.
+
+Including this type of field means that two subsequent runs of a build with
+identical inputs will produce different images.
+
+## Alternatives
+
+Alternatively, we could assert that Paketo buildpacks will not be making an
+effort to enable this type of reproducible build. However, this path doesn't
+seem reasonable. Supporting these types of reproducible builds is obviously
+possible given different choices that are not fundamental to the value
+proposition of buildpacks.
+
+## Implementation
+
+We should remove the `built_at` fields from all buildpack layers that include
+them. In our test suites that leverage this field to assert layer reuse, we
+should instead prefer to compare layer SHA values across rebuilds.
+
+We could also include new integration testing of this specific type of
+reproducible build scenario.
+
+## Unresolved Questions and Bikeshedding
+
+* How do we handle versioning of buildpacks that remove this layer metadata
+  field? Is this part of the public API?
```

Comment:
A couple of options come to mind:

- A feature flag to disable `built_at`. That way it remains, but someone can disable it and get reproducibility if they require it. You might be able to disable it by default if combined with some sort of deprecation period warning first.
- A deprecation period, followed by removal. We have some precedent for removing things following a deprecation period and not bumping a major version.
- Retain `build_at`, but always set it to 0. It would still exist, but the date value would always be the same. IMHO, that doesn't break the API cause the field still exists.

Some combination of these.

Reply:
Given that this particular type of layer metadata is intended for use by the buildpack itself (on rebuilds) and not by humans or other software, it would be convenient if we could treat this as an implementation detail and not a public API. On the other hand, the world abound with instances of people building dependencies on undocumented APIs especially when they are easily discoverable (like the metadata image label).

IMHO I think we should **not** treat this like a public API and instead document very clearly that the layer metadata is subject to change and should not be depended on. A couple reasons:
1. It seems unlikely that anyone is actually depending on this
2. It will be harder for us in the long term if we make guarantees about the metadata schemas
3. It is in keeping with our historical practice (I can think of examples where we changed layer metadata in incompatible ways in the Java buildpacks without cutting a major)
4. We didn't note this as a type of breaking change in the [semver RFC](https://github.com/paketo-buildpacks/rfcs/blob/main/text/0029-semantic-versioning.md#buildpack-general). Not sure that was deliberate vs we just didn't think about it but we could go back and clearly indicate that this type of change is patch level.